### PR TITLE
use `check_region_obligations_and_report_errors` to avoid ICEs

### LINKED ISSUE #53475: internal compiler error: `region_obligations not empty` with CoerceUnsized
`CoerceUnsized` into `dyn Any` causes an internal compiler panic

### Minimal Example:
```rust
#![feature(coerce_unsized)]

use std::ops::CoerceUnsized;
use std::any::Any;

struct Foo<T> {
	data: Box<T>
}

impl<T> CoerceUnsized<Foo<dyn Any>> for Foo<T> {}
```

This causes an internal compiler error:
```
$ cargo run
   Compiling rusting v0.1.0 (file:///tmp/rusting)                                                                                                                                      
error[E0601]: `main` function not found in crate `rusting`
  |
  = note: consider adding a `main` function to `src/main.rs`

thread 'main' panicked at 'region_obligations not empty: [
    (
        NodeId(
            17
        ),
        RegionObligation(sub_region=ReStatic, sup_type=T)
    )
]', librustc/infer/mod.rs:1036:9
note: Run with `RUST_BACKTRACE=1` for a backtrace.
error: aborting due to previous error

For more information about this error, try `rustc --explain E0601`.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.30.0-nightly (1fa944914 2018-08-17) running on x86_64-apple-darwin

note: compiler flags: -C debuginfo=2 -C incremental --crate-type bin

note: some of the compiler flags provided by cargo are hidden

error: Could not compile `rusting`.

To learn more, run the command again with --verbose.
```

This seems related to #51649, except this is using `CoerceUnsized`

## Meta
```$ rustc --version --verbose
rustc 1.30.0-nightly (1fa944914 2018-08-17)
binary: rustc
commit-hash: 1fa944914c092d728c8307e976a4b447df25bf16
commit-date: 2018-08-17
host: x86_64-apple-darwin
release: 1.30.0-nightly
LLVM version: 7.0
```

### PR DESCRIPTION
**Pull request by compiler-errors**
If we don't call `process_registered_region_obligations` before `resolve_regions_and_report_errors` then we'll ICE if we have any region obligations, and `check_region_obligations_and_report_errors` just does both of these for us in a nice convenient function.

Fixes #53475

r? types

### CODE REVIEW

File: compiler/rustc_typeck/src/coherence/builtin.rs [hunk 1]
```diff
@@ -606,7 +606,7 @@ pub fn coerce_unsized_info<'tcx>(tcx: TyCtxt<'tcx>, impl_did: DefId) -> CoerceUn
 
         // Finally, resolve all regions.
         let outlives_env = OutlivesEnvironment::new(param_env);
-        infcx.resolve_regions_and_report_errors(impl_did, &outlives_env);
+        infcx.check_region_obligations_and_report_errors(impl_did, &outlives_env);
```
Comment (compiler-errors) [inline compiler/rustc_typeck/src/coherence/builtin.rs:609]: This is what fixes #53475 --- the rest of them I haven't looked into whether they can be made to ICE similarly, but I don't see any harm in changing those too.

**Comment by oli-obk** [comment]
Are there any callers to `resolve_regions_and_report_errors` left? If so, maybe document this issue on the function itself to prevent additional mis-uses

@bors try @rust-timer queue

**Comment by oli-obk** [comment]
> Are there any callers to `resolve_regions_and_report_errors` left? If so, maybe document this issue on the function itself to prevent additional mis-uses

r=me with this resolved

@rustbot author

**Pull request merged**